### ISSUE TITLE
Standardised tracking events on the client

### ISSUE THREAD

**Issue by dmallory42** [Automattic/woocommerce-payments#8075]
### Description

This is a tech debt issue to tidy up the tracking logic on the client. Full scope to be determined, but this could include:

- Update the tracking code on the client to use TS so we can benefit from type hinting on props etc.
- update it to have a standardised place where all tracking events are stored as constants.

### Acceptance criteria

* Any tech debt relating to the tracking events on the client is taken care of.

**Comment by haszari** [comment]
Howdy @dmallory42 and thank you for tidying up our tracks code! 🙌🏻 

> update it to have a standardised place where all tracking events are stored as constants.

I don't think storing tracks event names in constants is worthwhile; I actually logged an issue for this (#8066) and coincidentally started fixing it for tracks events in my team's areas of responsibility (#8069). Unfortunately I didn't spot this issue / PR until now, thanks to @Jinksi for bringing it to my attention!

Would you mind taking a look at the rationale in #8066 and let me know your thoughts, specifically educate me on the benefits of using constants for event names? 

I did a quick check to see what other repos are doing, and I see inline event names in Jetpack, WooCommerce core and Calypso. I don't know of a reason why we'd have different needs, let me know what you think.

- Calypso (WPCOM): https://github.com/search?q=repo%3AAutomattic%2Fwp-calypso+recordTracksEvent&type=code&p=2
- Jetpack: https://github.com/search?q=repo%3AAutomattic%2Fjetpack%20recordEvent&type=code
- WooCommerce: https://github.com/search?q=repo%3Awoocommerce%2Fwoocommerce%20recordEvent&type=code

**Comment by dmallory42** [comment]
Hi @haszari! Thank you also for bringing the work your team are doing to my attention!

I understand your arguments about constants adding overhead, however I am of the opinion that using constants provide a variety of benefits for developers and should be considered as an approach. I totally agree that it would be great to reach an alignment on it. 

The main arguments towards using them are:
* **Readability and maintainability** - Using a centralised place to define constants makes it easier to manage tracking events. If we need to update an event name, we only need to do this in one place. The other benefit of this in my view is that if I, as a developer want to know which tracking events happen in our client, I can just look at the file where they are all defined and quickly see them, rather than needing to do a code search for `recordEvent` and sift through all the usages throughout our code base to find the event I need. This is also nice when working in an IDE because I can just type `events.` and then see an autocompleted list of all the events in our system, rather than needing to remember a specific string.
* **Consistency**:  I also believe using constants reduces the chance of making an error (e.g. typos) - I understand your view that these can be caught in PR review, but in my experience this doesn't always happen.

Regarding what other A8C codebases are doing, while I'm unsure of their rationale, I believe that may be at least partly historical (some of the codebases are 15+ years old) and probably the usage of constants added overhead, especially in systems where code is often developed on sandbox environments using vim, etc. I believe that with modern IDEs, using constants don't add such overhead. We also have the advantage of having Payments 'isolated' into a GH repo, rather than being part of a larger, monolithic application.

I do also see the other side of the argument however, and I think at least partly it comes down to personal developer preference. However it would be good to reach consensus on a consistent way of doing it (inline vs constants) - what do you think is the best way forward? Perhaps an RFC could be a good way to gauge how Transact devs feel about it? 

Thanks! 🙇

**Comment by ismaeldcom** [comment]
I'd like to share an intermediate solution that I mentioned in the PR review, I didn't push for it because we don't have a full TS codebase. However, now that I see both points of view, I think it may be the best option.

I personally don't like using constants to avoid string literals, but I like string literals even less when there is no tool to prevent us from making mistakes when writing them, it's too easy to have a typo.

With TypeScript in our codebase, we can take advantage of its string literal type to define each event under an `Event` type, allowing us to have a good coding experience with autocompletion and type checking, without the need for constants and their trade-offs.

Even if our code is a mix of JS and TS, we will be able to use a string in JS without any problems. The only drawback is that we can forget to add the event type when we call it from JS, but at least it is better than not having a list.

I think this issue is just a matter of taste, there is no right choice. Personally, I prefer to be able to check all tracked events in one place and not worry about typos. But that's just me 🙂 

Here is a quick example of what I mean:

<img width="593" alt="Screenshot 2024-02-08 at 11 53 44" src="https://github.com/Automattic/woocommerce-payments/assets/7670276/4b319f70-261c-47a6-a8db-0d062ee51ee7">

**Comment by haszari** [comment]
> I do also see the other side of the argument however, and I think at least partly it comes down to personal developer preference. However it would be good to reach consensus on a consistent way of doing it (inline vs constants) - what do you think is the best way forward? Perhaps an RFC could be a good way to gauge how Transact devs feel about it?

TLDR I think we should aim for consistency with others and I think it's important to inline them. 

I think Jetpack, Calypso, and WooCommerce are intentionally doing it this way for good reason. I've laid out the benefits in my issue here: #8066. As a user / consumer of tracks I think it's really important to consider the use-case of someone who needs to find out what events are fired, what they mean, by searching the code. Using constants is an impediment to that flow, for very little benefit to the developer (potentially preventing typos).

Tracks events are string constants that the developer should be free to iterate on. Adding typing to lock them down, or moving them to constants is overengineering in my view. If tracks event names are inline, consistent with other Automattic code bases, what's the worst thing that could happen? How big is the mistyping problem/risk?

I don't think we have special needs, and I think we should start with consistency with those code bases. If there's a good reason why we should go a different way (constants), then it should apply to those code bases too – do you think we should RFC to advocate for this change elsewhere? 

We could do an RFC – though is it a worthwhile use of everyone's time? 

> With TypeScript in our codebase, we can take advantage of its string literal type to define each event under an Event type, allowing us to have a good coding experience with autocompletion and type checking, without the need for constants and their trade-offs.

Thanks @ismaeldcom. This seems like a reasonable middle ground – at least this way the actual (true, real) event names are inline and searchable. This could work, though it does add an extra barrier to fixing/iterating on an event name.

**Comment by dmallory42** [comment]
Hi @haszari 👋 

> As a user / consumer of tracks I think it's really important to consider the use-case of someone who needs to find out what events are fired, what they mean, by searching the code. Using constants is an impediment to that flow, for very little benefit to the developer (potentially preventing typos).

I guess I didn't consider the use-case of someone needing to know what events are fired and what they mean - I was 
considering more the experience of day-to-day development - for me at least (I can only speak for myself as a developer), it's super useful to have a centralised way to see all the events in our system in one place and greatly reduces cognitive load - we also benefit from the IDE autocompletion I mentioned. I don't think it's as simple as just reducing typos - although that is also a plus.

> I don't think we have special needs, and I think we should start with consistency with those code bases. If there's a good reason why we should go a different way (constants), then it should apply to those code bases too – do you think we should RFC to advocate for this change elsewhere?

I agree we don't have special needs - but there are also numerous instances of us deciding to do things differently within Transact than how they are done in other parts of A8C because we think they are worthwhile. 

I'd also add that previously, we were using a mix of string constants and inline which I'm sure you agree isn't ideal, so whichever solution we decide on would be an improvement in terms of consistency. 

> We could do an RFC – though is it a worthwhile use of everyone's time?

No, probably not - but as I mentioned I think a large part of string constants vs inline is developer preference, so it may be the case that there is a consensus either way.

However, as I mentioned on Slack I'm happy to move back towards inline events - my preference would be to implement the solution @ismaeldcom laid out since I feel it gives us the best of both worlds, what do you think?

Thanks!

**Comment by haszari** [comment]
> However, as I mentioned on Slack I'm happy to move back towards inline events - my preference would be to implement the solution @ismaeldcom laid out since I feel it gives us the best of both worlds, what do you think?

Sounds like a great compromise, works for me! We get real event names inline ✅ and also an optional layer of typing protection as a developer convenience ✅ , e.g. for event names that are used more than once.

**Comment by dmallory42** [comment]
Great, thanks for taking the time to discuss it! I'll get a PR ready for that and tag you when it's in place in case you want to take a look!
